etc/afpd/directory.c

```diff
@@ -2094,6 +2418,29 @@ int setdirparams(struct vol *vol, struct path *path, uint16_t d_bitmap,
                 err = set_dir_errors(path, "setdirunixmode", errno);
             }
         }
+
+        /* Update dircache permissions to match filesystem after ACCESS or UNIXPR changes */
+        if ((set_maccess || set_upriv) && dir) {
+            struct stat st;
+
+            if (ostat(upath, &st, vol_syml_opt(vol)) == 0) {
+                /* Update cached ctime to match filesystem after our permission changes */
+                dir->dcache_ctime = st.st_ctime;
+
+                /* Recalculate rights cache based on new permissions */
+                if (dir == curdir) {
+                    /* Only update if dir is still curdir (hasn't changed during function) */
+                    /* Invalidate first to force accessmode() to recalculate from filesystem */
+                    dir->d_rights_cache = 0xffffffff;
+                    /* Use global AFPobj (always set during AFP session, see afp_dsi.c:498) */
```

Comment:
maybe you can refer to a particular function in that file rather than a line number?